### ISSUE TITLE
Possible bug on save model

### ISSUE THREAD

**Issue by sudongqi** [Element-Research/rnn#353]
This problem has bothered me for almost 2 weeks now. The loss from the training indicate the learning is happening, but whenever I load the pre-trained model using another script, the sampling from the model (seq2seq using SeqLSTM) doesn't make any sense to me. 

So I decided to do a test on an overfitting example tonight:

`require 'rnn'

file = torch.load('model.t7')
enc = file.enc
dec = file.dec
enc_lstm = file.enc_lstm
dec_lstm = file.dec_lstm
print ('load model')

--connection
function forwardConnect(enc, dec, len)
	for i=1,#enc_lstm do
		dec_lstm[i].userPrevOutput = enc_lstm[i].output[len]
		dec_lstm[i].userPrevCell = enc_lstm[i].cell[len]
	end
end
function backwardConnect(enc, dec)
	for i=1,#enc_lstm do
		enc_lstm[i].userNextGradCell = dec_lstm[i].userGradPrevCell
		enc_lstm[i].gradPrevOutput = dec_lstm[i].userGradPrevOutput
	end
end

--data
local encInSeq = torch.Tensor({{0,0,0,0,1,2,3},{0,0,0,4,3,2,1},{0,0,1,5,6,2,1}}):t()
local decInSeq = torch.Tensor({{6,1,2,3,4,0,0,0},{6,5,4,3,2,1,0,0},{6,3,4,3,2,0,0,0}}):t()
local decOutSeq = torch.Tensor({{1,2,3,4,7,0,0,0},{5,4,3,2,1,7,0,0},{3,4,3,2,7,0,0,0}}):t()

function predict(input_x)

	enc:evaluate()
   	dec:evaluate()

   	enc_lstm[1]:remember('eval')
   	dec_lstm[1]:remember('eval')

	input = torch.Tensor(1,7)
	input[1]=input_x
	input=input:t()

	local output = enc:forward(input)
	forwardConnect(enc,dec,input:size(1))

	local decInSeq = torch.Tensor({{6}})

	for i = 1, 20 do
		local p = dec:forward(decInSeq)[1]:t()
		local value, index = torch.max(p,1)
		local get = index[1][1]
		print(get)
		if get == 7 then break end
		decInSeq = torch.Tensor({{get}})
	end

	enc_lstm[1]:forget()
	dec_lstm[1]:forget()

	enc:forget()
	dec:forget()

	print ('\n')

end

--test
for i = 1,1000 do
	for j=1,3 do
		predict(encInSeq[{{},j}])
	end
end`

This sampling script returns nonsense even for an overfitting example. However, if I insert this piece of code right after the training script, the sampling work again.

`require 'rnn'

version = 1.3 -- Added multiple layers and merged with seqLSTM example

local opt = {}
opt.learningRate = 0.1
opt.hiddenSize = 6
opt.numLayers = 1
opt.useSeqLSTM = true -- faster implementation of LSTM + Sequencer
opt.vocabSize = 7
opt.seqLen = 7 -- length of the encoded sequence (with padding)
opt.niter = 100

enc_lstm={}
dec_lstm={}

function forwardConnect(enc, dec, len)
	for i=1,#enc_lstm do
		dec_lstm[i].userPrevOutput = enc_lstm[i].output[len]
		dec_lstm[i].userPrevCell = enc_lstm[i].cell[len]
	end
end
function backwardConnect(enc, dec)
	for i=1,#enc_lstm do
		enc_lstm[i].userNextGradCell = dec_lstm[i].userGradPrevCell
		enc_lstm[i].gradPrevOutput = dec_lstm[i].userGradPrevOutput
	end
end

-- Encoder
local enc = nn.Sequential()
enc:add(nn.LookupTableMaskZero(opt.vocabSize, opt.hiddenSize))
enc_lstm = {}
for i=1,opt.numLayers do
	enc_lstm[i] = nn.SeqLSTM(opt.hiddenSize, opt.hiddenSize)
	enc_lstm[i]:maskZero()
	enc:add(enc_lstm[i])

end
enc:add(nn.Select(1, -1))

-- Decoder
local dec = nn.Sequential()
dec:add(nn.LookupTableMaskZero(opt.vocabSize, opt.hiddenSize))
dec_lstm = {}
for i=1,opt.numLayers do
	dec_lstm[i] = nn.SeqLSTM(opt.hiddenSize, opt.hiddenSize)
	dec_lstm[i]:maskZero()
	dec:add(dec_lstm[i])
end
dec:add(nn.Sequencer(nn.MaskZero(nn.Linear(opt.hiddenSize, opt.vocabSize), 1)))
dec:add(nn.Sequencer(nn.MaskZero(nn.LogSoftMax(), 1)))
local criterion = nn.SequencerCriterion(nn.MaskZeroCriterion(nn.ClassNLLCriterion(),1))

--data
local encInSeq = torch.Tensor({{0,0,0,0,1,2,3},{0,0,0,4,3,2,1},{0,0,1,5,6,2,1}}):t()
local decInSeq = torch.Tensor({{6,1,2,3,4,0,0,0},{6,5,4,3,2,1,0,0},{6,3,4,3,2,0,0,0}}):t()
local decOutSeq = torch.Tensor({{1,2,3,4,7,0,0,0},{5,4,3,2,1,7,0,0},{3,4,3,2,7,0,0,0}}):t()

for i=1,1000 do

	enc:zeroGradParameters()
	dec:zeroGradParameters()

	-- Forward pass
	local encOut = enc:forward(encInSeq)
	forwardConnect(enc, dec, opt.seqLen)

	local decOut = dec:forward(decInSeq)
	--print(decOut)
	local err = criterion:forward(decOut, decOutSeq)
	
	print(string.format("Iteration %d ; NLL err = %f ", i, err))

	-- Backward pass
	local gradOutput = criterion:backward(decOut, decOutSeq)
	dec:backward(decInSeq, gradOutput)
	backwardConnect(enc, dec)
	local zeroTensor = torch.Tensor(encOut):zero()
	enc:backward(encInSeq, zeroTensor)

	dec:updateParameters(opt.learningRate)
	enc:updateParameters(opt.learningRate)
end

--save model
file={}
file.enc = nn.Serial(enc)
file.dec = nn.Serial(dec)

file.enc:mediumSerial()
file.dec:mediumSerial()

file.enc_lstm = enc_lstm
file.dec_lstm = dec_lstm
torch.save('model.t7',file)
print('save file')

function predict(input_x)

	enc:evaluate()
   	dec:evaluate()

   	enc_lstm[1]:remember('eval')
   	dec_lstm[1]:remember('eval')

	input = torch.Tensor(1,7)
	input[1]=input_x
	input=input:t()

	local output = enc:forward(input)
	forwardConnect(enc,dec,input:size(1))

	local decInSeq = torch.Tensor({{6}})

	for i = 1, 20 do
		local p = dec:forward(decInSeq)[1]:t()
		local value, index = torch.max(p,1)
		local get = index[1][1]
		print(get)
		if get == 7 then break end
		decInSeq = torch.Tensor({{get}})
	end

	enc_lstm[1]:forget()
	dec_lstm[1]:forget()

	enc:forget()
	dec:forget()

	print ('\n')

end

--test
for i = 1,1000 do
	for j=1,3 do
		predict(encInSeq[{{},j}])
	end
end
`

So it must be the error from the model saving function (or maybe the saving code I write is wrong? I have tried both Serial and the non-Serial option for saving.)
